pkg/apis/application/v1alpha1/types.go

```diff
@@ -1523,14 +1573,93 @@ func (w *ProjectMaintenanceWindows) Match(app *Application) (bool, ProjectMainte
 			}
 		}
 		if len(matchingWindows) > 0 {
-			return true, matchingWindows
+			return &matchingWindows
 		}
 	}
-	// No active maintenance windows
-	return false, nil
+	return nil
+}
+
+func (w *SyncWindows) CanAutoSync() bool {
+	var allow bool
+	if w.HasWindows() {
+		var denyActive bool
+		active := w.Active()
+		if active.HasWindows() {
+			for _, a := range *active {
+				if a.Kind == "deny" {
+					allow = false
+					denyActive = true
+				}
+				if a.Kind == "allow" {
+					if !denyActive {
+						allow = true
+					}
+				}
+			}
+		}
+		if !w.InactiveAllows().HasWindows() && !denyActive {
+			allow = true
+		}
+	} else {
+		allow = true
+	}
+	return allow
+}
+
+func (w *SyncWindows) CanSync() bool {
```

Comment:
Just a suggestion. You can return true as of `HasWindows` is false. This would allow to avoid nested if block:

```
if !w.HasWindows() {
  return true
}
...
```